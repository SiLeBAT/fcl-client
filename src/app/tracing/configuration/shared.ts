import {
    DeliveryHighlightingRule, HighlightingRule, LinePatternType,
    StationHighlightingRule, TableColumn, TableRow
} from '../data.model';
import { PropToValuesMap } from './configuration.model';
import {
    ColorAndShapeEditRule, ColorEditRule, InvEditRule,
    LabelEditRule, RowFilter, RuleType, SimpleLabelEditRule
} from './model';
import * as _ from 'lodash';

export type EditRuleOfType<T extends RuleType> =
    T extends RuleType.LABEL ? LabelEditRule :
    T extends RuleType.COLOR_AND_SHAPE ? ColorAndShapeEditRule :
    T extends RuleType.COLOR ? ColorEditRule :
    T extends RuleType.INVISIBILITY ? InvEditRule : never;

export function filterTableRows(rows: TableRow[], rowFilters: RowFilter<any>[]): TableRow[] {
    let filteredRows = rows;
    for (const rowFilter of rowFilters) {
        filteredRows = rowFilter.filter(filteredRows);
    }
    return filteredRows;
}

export function extractPropToValuesMap(tableRows: TableRow[], tableColumns: TableColumn[]): PropToValuesMap {
    const propToValuesMap: PropToValuesMap = {};
    for (const column of tableColumns) {
        const values: string[] = _.uniq(
            tableRows
                .map(r => r[column.id] as (string | number | boolean))
                .filter(v => v !== undefined && v !== null)
                .map(v => typeof v === 'string' ? v : '' + v)
        ).sort();
        propToValuesMap[column.id] = values;
    }

    propToValuesMap[''] = _.uniq([].concat(...Object.values(propToValuesMap))).sort();

    return propToValuesMap;
}

export function isSimpleLabelRule(editRule: LabelEditRule): editRule is SimpleLabelEditRule {
    return (editRule as SimpleLabelEditRule).labelProperty !== undefined;
}

export function createDefaultHRule(): Omit<HighlightingRule, 'id' | 'name'> {
    return {
        userDisabled: false,
        autoDisabled: false,
        showInLegend: false,
        color: null,
        invisible: false,
        adjustThickness: false,
        labelProperty: null,
        valueCondition: null,
        logicalConditions: [[]]
    };
}

export function createDefaultStatHRule(): Omit<StationHighlightingRule, 'id' | 'name'> {
    return {
        ...createDefaultHRule(),
        shape: null
    };
}

export function createDefaultDelHRule(): Omit<DeliveryHighlightingRule, 'id' | 'name'> {
    return {
        ...createDefaultHRule(),
        linePattern: LinePatternType.SOLID
    };
}
