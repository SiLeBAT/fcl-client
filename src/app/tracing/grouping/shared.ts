import {LinkGroup, GroupInfoFun} from './model';
import {GroupData} from '../data.model';
import * as _ from 'lodash';

export function extractNewGroups(
  map: Map<string, LinkGroup>,
  getGroupInfo: GroupInfoFun
): GroupData[] {
  const compareNumbers = (a, b) => (a < b ? -1 : a === b ? 0 : 1);
  const newGroups: GroupData[] = [];

  map.forEach(linkStation => {
    linkStation.linkedStations.sort((a, b) =>
      compareNumbers(a.linkKeys.length, b.linkKeys.length)
    );

    let size = 0;
    const linkIndices: Set<number> = new Set();
    // eslint-disable-next-line one-var
    for (
      let iLS = 0, nLS: number = linkStation.linkedStations.length;
      iLS < nLS;
      iLS++
    ) {
      if (linkStation.linkedStations[iLS].linkKeys.length !== size) {
        addNewGroups(linkStation, newGroups, linkIndices, getGroupInfo);
        size = linkStation.linkedStations[iLS].linkKeys.length;
      }
      linkIndices.add(iLS);
    }
    addNewGroups(linkStation, newGroups, linkIndices, getGroupInfo);
  });
  return newGroups;
}

export function addNewGroups(
  linkGroup: LinkGroup,
  newGroups: GroupData[],
  linkIndices: Set<number>,
  getGroupInfo: GroupInfoFun
) {
  if (linkIndices.size > 1) {
    let groupMembers: string[] = [];
    while (linkIndices.size > 0) {
      const compareIndex: number = linkIndices.values().next().value;
      linkIndices.delete(compareIndex);

      const compareKeys: string[] =
        linkGroup.linkedStations[compareIndex].linkKeys;
      groupMembers.push(
        linkGroup.linkedStations[compareIndex].linkedStation.id
      );

      const removeIndices: number[] = [];
      linkIndices.forEach(i => {
        if (_.isEqual(compareKeys, linkGroup.linkedStations[i].linkKeys)) {
          removeIndices.push(i);
        }
      });
      removeIndices.forEach(i => {
        linkIndices.delete(i);
        groupMembers.push(linkGroup.linkedStations[i].linkedStation.id);
      });

      if (groupMembers.length > 1) {
        newGroups.push({
          ...getGroupInfo(linkGroup, newGroups.length + 1),
          contains: groupMembers,
        });
      }
      groupMembers = [];
    }
  }
  linkIndices.clear();
}
