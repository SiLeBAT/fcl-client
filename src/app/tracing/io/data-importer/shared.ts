import Ajv from 'ajv';
import { HighlightingSettings, OperationType, ValueType, LinePatternType } from '../../data.model';
import { InputFormatError } from '../io-errors';

const STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX = 'SDHR';
const DELIVERY_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX = 'DDHR';

export async function isValidJson(schema: any, data: any, throwError?: boolean): Promise<boolean> {
    const ajv = new Ajv();
    const valid = ajv.validate(schema, data);
    if (!valid && throwError) {
        throw new InputFormatError('Invalid json schema: ' + ajv.errorsText(ajv.errors));
    }
    return valid;
}

export function compareVersions(version1: string, version2: string): number {
    const versionNumbers1: Number[] = version1.split('.').map(s => Number(s));
    const versionNumbers2: Number[] = version2.split('.').map(s => Number(s));

    for (let i = 0; i < 3; i++) {
        if (versionNumbers1[i] !== versionNumbers2[i]) {
            return versionNumbers1[i] < versionNumbers2[i] ? -1 : 1;
        }
    }
    return 0;
}

export function areMajorVersionsMatching(version1: string, version2: string): boolean {
    const versionNumbers1: Number[] = version1.split('.').map(s => Number(s));
    const versionNumbers2: Number[] = version2.split('.').map(s => Number(s));

    return versionNumbers1[0] === versionNumbers2[0];
}

export function checkVersionFormat(version: String): boolean {
    return version && version.trim().match('^\\d+\\.\\d+\\.\\d+$').length > 0;
}

export function createDefaultHighlights(): HighlightingSettings {
    const defaultHighlights: HighlightingSettings = {
        invisibleStations: [],
        invisibleDeliveries: [],
        stations: [
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Outbreak',
                name: 'Outbreak',
                showInLegend: true,
                disabled: false,
                color: [
                    255,
                    0,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'weight',
                            operationType: OperationType.GREATER,
                            value: '0'
                        }
                    ]
                ],
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Observed',
                name: 'Observed',
                showInLegend: true,
                disabled: false,
                color: [
                    0,
                    255,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'observed',
                            operationType: OperationType.NOT_EQUAL,
                            value: 'none'
                        }
                    ]
                ],
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Forward Trace',
                name: 'Forward Trace',
                showInLegend: true,
                disabled: false,
                color: [
                    255,
                    200,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'forward',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Backward Trace',
                name: 'Backward Trace',
                showInLegend: true,
                disabled: false,
                color: [
                    255,
                    0,
                    255
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'backward',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Cross Contamination',
                name: 'Cross Contamination',
                showInLegend: true,
                disabled: false,
                color: [
                    0,
                    0,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'crossContamination',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Common Link',
                name: 'Common Link',
                showInLegend: true,
                disabled: false,
                color: [
                    255,
                    255,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'score',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Score',
                name: 'Score',
                showInLegend: false,
                disabled: false,
                color: null,
                invisible: false,
                adjustThickness: true,
                labelProperty: null,
                valueCondition: {
                    propertyName: 'score',
                    valueType: ValueType.VALUE,
                    useZeroAsMinimum: true
                },
                logicalConditions: null,
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'StationLabel',
                name: 'StationLabel',
                showInLegend: false,
                disabled: false,
                color: null,
                invisible: false,
                adjustThickness: false,
                labelProperty: 'name',
                valueCondition: null,
                logicalConditions: [
                    []
                ],
                shape: null
            },
            {
                id: STATION_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Kill Contamination',
                name: 'Kill Contamination',
                showInLegend: true,
                disabled: false,
                color: [ 153, 153, 153 ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'killContamination',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                shape: null
            }
        ],

        deliveries: [
            {
                id: DELIVERY_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Outbreak',
                name: 'Outbreak',
                showInLegend: true,
                disabled: false,
                color: [
                    255,
                    0,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'weight',
                            operationType: OperationType.GREATER,
                            value: '0'
                        }
                    ]
                ],
                linePattern: LinePatternType.SOLID
            },
            {
                id: DELIVERY_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Observed',
                name: 'Observed',
                showInLegend: true,
                disabled: false,
                color: [
                    0,
                    255,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'observed',
                            operationType: OperationType.NOT_EQUAL,
                            value: 'none'
                        }
                    ]
                ],
                linePattern: LinePatternType.SOLID
            },
            {
                id: DELIVERY_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Forward Trace',
                name: 'Forward Trace',
                showInLegend: true,
                disabled: false,
                color: [
                    255,
                    200,
                    0
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'forward',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                linePattern: LinePatternType.SOLID
            },
            {
                id: DELIVERY_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Backward Trace',
                name: 'Backward Trace',
                showInLegend: true,
                disabled: false,
                color: [
                    255,
                    0,
                    255
                ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'backward',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                linePattern: LinePatternType.SOLID
            },
            {
                id: DELIVERY_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'Kill Contamination',
                name: 'Kill Contamination',
                showInLegend: true,
                disabled: false,
                color: [ 153, 153, 153 ],
                invisible: false,
                adjustThickness: false,
                labelProperty: null,
                valueCondition: null,
                logicalConditions: [
                    [
                        {
                            propertyName: 'killContamination',
                            operationType: OperationType.EQUAL,
                            value: '1'
                        }
                    ]
                ],
                linePattern: LinePatternType.SOLID
            },
            {
                id: DELIVERY_DEFAULT_HIGHLIGHTING_RULE_ID_PREFIX + 'DeliveryLabel',
                name: 'DeliveryLabel',
                showInLegend: false,
                disabled: false,
                color: null,
                invisible: false,
                adjustThickness: false,
                labelProperty: 'name',
                valueCondition: null,
                logicalConditions: [
                    []
                ],
                linePattern: LinePatternType.SOLID
            }
        ]

    };

    return defaultHighlights;
}
