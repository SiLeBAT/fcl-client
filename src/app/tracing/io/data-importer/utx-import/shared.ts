import { PropertyEntry } from "@app/tracing/data.model";
import { Registration, RegistrationScheme } from "./utx-model";

type JSON_VALUE = string | number | boolean | object;
type SIMPLE_TYPE = string | number | boolean;

export function getRegistration(
    registrations: Registration[],
    registrationSchemeMap: Map<string, RegistrationScheme>,
):
    | {
          number: string | undefined;
          type: string | undefined;
          registry: string | undefined;
      }
    | undefined {
    if (registrations.length === 0) {
        return undefined;
    }
    const referenceRegistration = registrations[0];
    return {
        number: referenceRegistration.registrationNumber,
        type: referenceRegistration.registrationType,
        registry: registrationSchemeMap.get(
            referenceRegistration.registrationScheme ?? "",
        )?.name,
    };
}

export function mergeCVAndTreeText(
    cvValue: string | undefined,
    freeTextValue: string | undefined,
    otherValue: string,
): string | undefined {
    if (cvValue === otherValue) {
        return freeTextValue;
    }
    return cvValue;
}

export function mergeCVsAndTreeTexts(
    cvValues: string[] | undefined,
    freeTextValues: string[] | undefined,
    otherValue: string,
): string[] | undefined {
    if (cvValues === undefined) {
        return undefined;
    }
    const definedFreeTextValues = freeTextValues || [];
    let freeTextIndex = 0;
    const mergedValues = cvValues.map((cvValue) =>
        cvValue === otherValue
            ? (definedFreeTextValues[freeTextIndex++] ?? "")
            : cvValue,
    );
    return mergedValues;
}

export function value2SimpleType(
    value: JSON_VALUE | undefined,
): SIMPLE_TYPE | undefined {
    if (typeof value === "object") {
        if (Array.isArray(value) && value.length > 0) {
            return value[0];
        }
        return undefined;
    }
    return value;
}

export function createProperties(
    entries: { id: string; value: JSON_VALUE | undefined }[],
): PropertyEntry[] {
    const properties = entries.map((e) => ({
        name: e.id,
        value: value2SimpleType(e.value),
    }));
    const propertiesWithValues = properties.filter(
        (p) => p.value !== undefined,
    ) as PropertyEntry[];
    return propertiesWithValues;
}
