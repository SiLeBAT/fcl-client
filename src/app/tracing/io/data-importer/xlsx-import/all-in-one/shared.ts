export class Register {
    private map = new Map<string, number>();

    has(id: string): boolean {
        return this.map.has(id);
    }

    private getCount(id: string): number {
        return this.map.get(id) ?? 0;
    }

    add(id: string): void {
        this.map.set(id, this.getCount(id) + 1);
    }

    isRegisteredOnce(id: string): boolean {
        return this.getCount(id) === 1;
    }
}
