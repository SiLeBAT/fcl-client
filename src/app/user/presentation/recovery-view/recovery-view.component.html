<div class="fcl-image-container">
    <fcl-main-entry-header></fcl-main-entry-header>

    <fcl-single-center-card-layout
        class="fcl-register-card"
        cardtitle="Recover Password"
    >
        <fcl-mat-card-body>
            <fcl-recovery-container></fcl-recovery-container>
        </fcl-mat-card-body>
        <fcl-mat-card-actions> </fcl-mat-card-actions>
    </fcl-single-center-card-layout>
</div>
