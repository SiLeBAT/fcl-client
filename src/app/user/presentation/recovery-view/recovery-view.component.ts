import {Component} from '@angular/core';

@Component({
  selector: 'fcl-recovery-view',
  templateUrl: './recovery-view.component.html',
  styleUrls: ['./recovery-view.component.scss'],
})
export class RecoveryViewComponent {}
