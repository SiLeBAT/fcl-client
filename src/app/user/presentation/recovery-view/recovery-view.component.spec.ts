import { async, ComponentFixture, TestBed } from "@angular/core/testing";
import { CUSTOM_ELEMENTS_SCHEMA } from "@angular/core";

import { RecoveryViewComponent } from "./recovery-view.component";
import { MaterialModule } from "../../../shared/material.module";
import { SharedModule } from "../../../shared/shared.module";
import { RouterTestingModule } from "@angular/router/testing";

describe("RecoveryViewComponent", () => {
    let component: RecoveryViewComponent;
    let fixture: ComponentFixture<RecoveryViewComponent>;

    beforeEach(async(() => {
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        TestBed.configureTestingModule({
            declarations: [RecoveryViewComponent],
            imports: [MaterialModule, SharedModule, RouterTestingModule],
            schemas: [CUSTOM_ELEMENTS_SCHEMA],
        })
            .compileComponents()
            .then(() => {
                fixture = TestBed.createComponent(RecoveryViewComponent);
                component = fixture.componentInstance;
            });
    }));

    it("should create the component", () => {
        expect(component).toBeTruthy();
    });

    it("should render the component correctly", () => {
        expect(fixture).toMatchSnapshot();
    });
});
