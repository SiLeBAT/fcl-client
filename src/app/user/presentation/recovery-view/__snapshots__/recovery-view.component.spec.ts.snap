// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecoveryViewComponent should render the component correctly 1`] = `
<fcl-recovery-view>
  <div
    class="fcl-image-container"
  >
    <fcl-main-entry-header />
    <fcl-single-center-card-layout
      cardtitle="Recover Password"
      class="fcl-register-card"
    >
      <div
        class="fcl-single-center-card-layout-item fcl-single-center-card-layout-container"
      >
        <mat-card
          class="mat-card mat-focus-indicator fcl-mat-card"
        >
          <mat-card-header
            class="mat-card-header fcl-mat-card__header"
          >
            <div
              class="mat-card-header-text"
            >
              <mat-card-title
                class="mat-card-title"
              >
                <div
                  class="mat-headline"
                >
                </div>
                <mat-divider
                  class="mat-divider"
                  role="separator"
                />
              </mat-card-title>
            </div>
          </mat-card-header>
          <mat-card-content
            class="mat-card-content"
          >
            <fcl-mat-card-body>
              <fcl-recovery-container />
            </fcl-mat-card-body>
          </mat-card-content>
          <mat-card-actions
            class="mat-card-actions fcl-mat-card-actions"
          >
            <fcl-mat-card-actions />
          </mat-card-actions>
        </mat-card>
      </div>
    </fcl-single-center-card-layout>
  </div>
</fcl-recovery-view>
`;
